### PR TITLE
Split `internal.d.ts` into multiple files

### PR DESCRIPTION
**Pull request by tommy-mitchell**
Splits the types in `internal.d.ts` into multiple category-based files to improve maintainability, with the following structure:

```
\_ source/
   \_ internal/
      \_ array.d.ts
      \_ characters.d.ts
      \_ keys.d.ts
      \_ numeric.d.ts
      \_ object.d.ts
      \_ string.d.ts
      \_ tuple.d.ts
      \_ type.d.ts (misc types)
      \_ index.d.ts (so imports don't need updated)
   \_ ...
\_ test-d/
   \_ internal/
      \_ [category]/
         \_ [type].ts
   \_ ...
```

Other miscellaneous changes:

- Updates test script to detect source files in subfolders.
- Removes unused type imports.
- Renames `ArrayMax`/`ArrayMin` to `TupleMax`/`TupleMin`. This can be reverted if desired.

I tried maintaining the order of types and their formatting when copying them over, but I could've missed something.

### CODE REVIEW

**Comment by sindresorhus** [comment]
I don't mind this, but I also don't see a strong need to do it. Ideally, many of these types should be exported publicly instead (https://github.com/sindresorhus/type-fest/issues/676).

One thing I don't want is to nest the tests into folder. I think all the internal types should be top-level tests in the `tests/internal` folder.

**Comment by tommy-mitchell** [comment]
Updated tests and exposes some internal types that I found requests for in the issues. Please see the lines marked `TODO` for outstanding questions.

**Comment by sindresorhus** [comment]
When I mentioned exporting types, I didn't really mean in this pull request. It would be better to have separate pull requests for each exposed type, to make it easier to review and also to track changes for release notes.

File: readme.md [hunk 1]
```diff
@@ -194,13 +194,16 @@ Click the type names for complete docs.
...
+- [`ArrayElement`](source/array-element.d.ts) - Extracts the element type of an array or union of arrays.
+- [`ArrayTail`](source/array-tail.d.ts) - Extracts the type of an array or tuple minus the first element.
 - [`SetFieldType`](source/set-field-type.d.ts) - Create a type that changes the type of the given keys.
 - [`Paths`](source/paths.d.ts) - Generate a union of all possible paths to properties in the given object.
 - [`SharedUnionFieldsDeep`](source/shared-union-fields-deep.d.ts) - Create a type with shared fields from a union of object types, deeply traversing nested structures.
 - [`DistributedOmit`](source/distributed-omit.d.ts) - Omits keys from a type, distributing the operation over a union.
 - [`DistributedPick`](source/distributed-pick.d.ts) - Picks keys from a type, distributing the operation over a union.
 - [`And`](source/and.d.ts) - Returns a boolean for whether two given types are both true.
 - [`Or`](source/or.d.ts) - Returns a boolean for whether either of two given types are true.
+- [`NonEmptyTuple`](source/non-empty-tuple.d.ts) - Matches any non empty tuple.
```
Comment (sindresorhus) [inline readme.md:206]:
```suggestion
- [`NonEmptyTuple`](source/non-empty-tuple.d.ts) - Matches any non-empty tuple.
```

File: source/non-empty-tuple.d.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+//=> 6
+
+sum();
+//=> Error: Expected at least 1 arguments, but got 0.
+```
+
+@see {@link RequireAtLeastOne} for objects
+
+@category Tuple
+// TODO: array?
```
Comment (sindresorhus) [inline source/non-empty-tuple.d.ts:20]: `Array` only

File: source/non-empty-tuple.d.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
+/**
+Matches any non empty tuple.
```
Comment (sindresorhus) [inline source/non-empty-tuple.d.ts:2]:
```suggestion
Matches any non-empty tuple.
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by tommy-mitchell** [comment]
Unexposed types and added to separate PRs.

File: source/internal/array.d.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+@link https://itnext.io/implementing-arithmetic-within-typescripts-type-system-a1ef140a6f6f
+*/
+type ArrayLength<T extends readonly unknown[]> = T extends {readonly length: infer L} ? L : never;
+
+/**
+Matches any unknown array or tuple.
+*/
+export type UnknownArrayOrTuple = readonly [...unknown[]];
+// TODO: should unknown-array be updated?
```
Comment (tommy-mitchell) [inline source/internal/array.d.ts:14]: Do we need both of these types?

Comment (sindresorhus) [reply source/internal/array.d.ts:14]: I don't know. I don't think that's relevant to this pull request though. Open a separate issue about it.

**Pull request merged**